### PR TITLE
[Enhancement] low-cardinality intersect add distinct

### PR DESCRIPTION
**Pull request by Seaven**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Tpcds1t query14 12s -> 8s

Do distinct before executing intersect when input rows are low cardinality

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
starrocks_be_unittest succeeded.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 33 / 33 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSet.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleSetType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java|29|29|100.00%|[]|

**Review by kangkaisen — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+import com.starrocks.sql.optimizer.operator.pattern.Pattern;
+import com.starrocks.sql.optimizer.rule.RuleType;
+import com.starrocks.sql.optimizer.statistics.StatisticsCalculator;
+
+import java.util.List;
+
+/*
+ *                                    Intersect
+ *       Intersect                  /     |     \
+ *     /     |     \       -->  Distinct  |   Distinct
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java:22]: Node2 misses Distinct?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+        LogicalAggregationOperator agg = new LogicalAggregationOperator(AggType.GLOBAL,
+                Lists.newArrayList(intersect.getChildOutputColumns().get(child)), Maps.newHashMap());
+
+        OptExpression aggOpt = OptExpression.create(agg, childOpt);
+        calculateStatistics(aggOpt, context);
+        double originRows = childOpt.getStatistics().getOutputRowCount();
+        double aggRows = aggOpt.getStatistics().getOutputRowCount();
+
+        if (aggRows <= 0 || originRows <= 0 || aggRows * 1000 > originRows) {
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java:55]: What 1000 means

Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java:55]: means it's a low cardinality aggregation, I will replace all magic number to constant variable in next PR

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java (hunk 1 above)
Comment (Seaven) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/IntersectAddDistinctRule.java:22]: Node2 maybe a high cardinality aggregation, don't need add distinct

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**